Documentation/teaching/labs/networking.rst

```diff
@@ -0,0 +1,1251 @@
+Endian <http://en.wikipedia.org/wiki/Endianness#Little-endian>`_ (the least
+significant byte first). Since a network interconnects systems with different
+platforms, the Internet has imposed a standard sequence for the storage of
+numerical data, called `network byte-order
+<http://en.wikipedia.org/wiki/Endianness#Endianness_in_networking>`_. In
+contrast, the byte sequence for the representation of numerical data on the host
+computer is called host byte-order. Data received/sent from/to the network is in
+the network byte-order format and should be converted between this format and
+the host byte-order.
+
+For converting we use the following macros:
+
+  * ``u16 htons(u16 x)`` converts a 16 bit integer from host byte-order to
+    network byte-order (host to network short);
+  * ``u32 htonl(u32 x)`` converts a 32 bit integer from host byte-order to
+    network byte-order (host to network long);
+  * ``u16 ntohs(u16 x)`` converts a 16 bit integer from network byte-order to
+    host byte-order (network to host short);
+  * ``u32 ntohl(u32 x)`` converts a 32 bit integer from network byte-order to
+    host byte-order (network to host long).
+
+.. _netfilter:
+
+netfilter
+=========
+
+Netfilter is the name of the kernel interface for capturing network packets for
+modifying/analyzing them (for filtering, NAT, etc.). `The netfilter
+<http://www.netfilter.org/>`_ interface is used in user-space by `iptables
+<http://www.frozentux.net/documents/iptables-tutorial/>`_.
+
+In the Linux kernel, packet capture using netfilter is done by attaching hooks.
+Hooks can be specified in different locations in the path followed by a kernel
+network packet, as needed. An organization chart with the route followed by a
+package and the possible areas for a hook can be found here.
+
+The header included when using netfilter is :file:`linux/netfilter.h`.
+
+A hook is defined through the :c:type:`struct nf_hook_ops` structure:
+
+.. code-block:: c
+
+  struct nf_hook_ops {
+  	/* User fills in from here down. */
+  	nf_hookfn               *hook;
+  	struct net_device       *dev;
+  	void                    *priv;
+  	u_int8_t                pf;
+  	unsigned int            hooknum;
+  	/* Hooks are ordered in ascending priority. */
+  	int                     priority;
+  };
+
+where:
+
+  * ``pf`` is the package type (``PF_INET``, etc.);
+  * ``priority`` is the priority; priorities are defined in
+     :file:`uapi/linux/netfilter_ipv4.h` as follows:
+
+.. code-block:: c
+
+  enum nf_ip_hook_priorities {
+  	NF_IP_PRI_FIRST = INT_MIN,
+  	NF_IP_PRI_CONNTRACK_DEFRAG = -400,
+  	NF_IP_PRI_RAW = -300,
+  	NF_IP_PRI_SELINUX_FIRST = -225,
+  	NF_IP_PRI_CONNTRACK = -200,
+  	NF_IP_PRI_MANGLE = -150,
+  	NF_IP_PRI_NAT_DST = -100,
+  	NF_IP_PRI_FILTER = 0,
+  	NF_IP_PRI_SECURITY = 50,
+  	NF_IP_PRI_NAT_SRC = 100,
+  	NF_IP_PRI_SELINUX_LAST = 225,
+  	NF_IP_PRI_CONNTRACK_HELPER = 300,
+  	NF_IP_PRI_CONNTRACK_CONFIRM = INT_MAX,
+  	NF_IP_PRI_LAST = INT_MAX,
+  };
+
+\
```

Comment:
Remove ``\``

Reply:
I left it there intentionally. I wanted to keep the indentation from above the code snippet. If I didn't have this, the paragraph would be included in the code snippet.